### PR TITLE
Create 0004-releases.md

### PR DESCRIPTION
**Pull request by digitalsadhu**
Rendered version: https://github.com/fabric-ds/issues/blob/403703b7bb5390e31b52bd479aab1133350e30b7/rfcs/0004-releases.md

### CODE REVIEW

**Comment by pearofducks** [comment]
Are breaking changes the only ones considered here?

If not, would we delay getting a bugfix out?
Likewise, with other non-breaking changes (e.g. UI tweaks), we would delay those?

**Comment by digitalsadhu** [comment]
> Are breaking changes the only ones considered here?
> 
> If not, would we delay getting a bugfix out? Likewise, with other non-breaking changes (e.g. UI tweaks), we would delay those?

Forgot to include anything on this but yes, everything included but any important bug fixes would be back ported via a cherry pick to main is my thinking. Most likely, this would be non breaking bug fixes so upgrade guides would not be necessary but automated changelog and slack messaging would still happen.

File: rfcs/0004-releases.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+We can't allow a fixed release cycle to block us from getting immediate work into the hands of those who can't wait weeks for the next release.
+To solve this we should develop against a `next` branch that is constantly published to NPM and Eik using a `next` tag. (eg. `1.0.0-next.1`)
+Anyone needing a feature we have only just completed can install the `next` version while waiting for the next release window.
+Cutting an actual release then would simply be a matter of merging `next` into `main` and writing an upgrade guide.
+
+
+### Backport bugfixes and urgent features
+
+Anything that must go out on the main channel can as agreed upon by the team. To do so, we can cherry pick commits from next to main. Changelogs and slack notifications will go out and might include an explanation as to why we felt it was important enough to not wait for the next release cycle. Upgrade guides would typically not be required since such changes will likely not include new features or breaking changes. If for some reason they do, an upgrade guide would be necessary.
```
Comment (BalbinaK) [inline rfcs/0004-releases.md:41]: Something sounds off in the first sentence here :)

**Comment by digitalsadhu** [comment]
Perhaps, but I think we should be able to balance those scales pretty well though. It should be possible to automate for the most part such that:

1. We always write good semantic commit messages
1. We PR against and then merge to `next` during day to day development
1. Once per release cycle we merge to main and write an upgrade guide

That really should be it I think. The rest can be automated. With the caveat that every so often we may need to cherry pick a fix from next to main if it's super urgent.

Most of the effort should be in getting the setup in place which is essentially a one off cost.

The benefits then, in my mind, start to outweigh the costs since once in place, you have a regular, predictable release schedule that teams can plan their work around with good documentation that makes it easy to make upgrades. 

This type of approach should raise the bar somewhat and is something we've been discussing on FPW for a while. We are looking to implement something like this in Podium, Eik etc. as its one of the areas we've identified for improvement as part of the new platform initiatives. (pushing more and more toward "self service" for teams that use our products)

File: rfcs/0004-releases.md (hunk 1 above)
Comment (digitalsadhu) [reply rfcs/0004-releases.md:41]: ah yes... indeed!

Comment (digitalsadhu) [reply rfcs/0004-releases.md]: Fixed!

**Comment by digitalsadhu** [comment]
Another point to note is that due to the way Eik and aliasing strategy focuses on small non breaking changes overtime in order to line apps and teams up on the same versions of globally shared libraries like Fabric, there is a small inherent risk that we could accidentally roll out bad code to a lot of people without them explicitly making an update themselves. 

We counter this risk as best we can in a couple ways: 

1. Its possible to rollback
2. We draw a line between breaking (major) changes and all other changes and expect users to manually update across breaking changes. ie. We will never automatically roll out a major change to anyone. 

However, there is a 3rd way we would like to add to help with this which is to give people a window within which they have a chance to test against upcoming changes. This comes down a lot to communication and it fits in well with a fixed regular release cycle. 

For example:
If we were to release say once per month, we could also always roll out the equivalent Eik update 1 week later so that everyone has a chance to test their apps against the latest release before it goes live. This could even be in the form of an automated smoke test.

**Comment by trygve-lie** [comment]
Good write up.

For me one of the most important parts here is to automate the release cycle. From experience we see that its just not possible to provide good release standards when its done manually. Automate it helps a lot.

I am +1 for such a work flow as long as we automate the whole release process. 

Though, on the topic of a time slot for cutting releases one need to find the sweet spot between it being too long and too short between releases plus it must be fully possible to cut releases outside of this time slot. I don't see anything preventing that in this suggestion.

> Maybe I'm missing something, but I don't understand when 99% of the time we aren't breaking anything and there's no implied upgrade effort we want to add effort to our deployments.

Keeping a good and high test coverage safeguards against breaking things but being able to cut `next` releases before a full release is an extra level of protection in my mind. With such a release cycle we are in a position to constantly push `next` releases for each minor change and then we are able to let selected apps test against the `next` release in real apps before full releases is cut. That is a pretty big advantage.

**Comment by digitalsadhu** [comment]
Any further comments/thoughts/questions from anyone? @pearofducks Do you still have doubts you feel we should resolve?

Next step once we have agreement is to merge this PR and create a ticket to create/test out a reference implementation

**Comment by pearofducks** [comment]
All good for me, thanks for checking!

**Pull request merged**